KAFKA-18698: Migrate suitable classes to records in server and server-common modules

frankvicky opened the pull request:
JIRA: KAFKA-18698
The following are the classes that might be converted to records:
```
server-common:
TopicIdPartition
RequestAndCompletionHandler
FinalizedFeatures

server:
Assignment
```

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

frankvicky committed:
KAFKA-18698: Migrate suitable classes to records in server and server-common modules

JIRA: KAFKA-18698
The following are the classes that might be converted to records:
```
server-common:
TopicIdPartition
RequestAndCompletionHandler
FinalizedFeatures

server:
Assignment
```

frankvicky commented on server-common/src/main/java/org/apache/kafka/server/common/TopicIdPartition.java:
```diff
@@ -18,45 +18,15 @@
+    Uuid topicId,
+    int partitionId
+) {
     @Override
     public String toString() {
         return topicId + ":" + partitionId;
```
`org.apache.kafka.server.AssignmentTest#testAssignmentToString` has an assertion for the `toString` output. Hence, I don't remove this one.

clolov replied:
Maybe I am missing something obvious, but why can't you just change the output of the test to adhere to the new format and use the new format?

m1a2st replied:
In `org.apache.kafka.server` there is `package-info.java`, I think it is public api, thus we should not modify the `toString` method

clolov replied:
I understand, so the cause to keep it as is is that it is public and not because a test is using it, okay

chia7712 replied:
> In org.apache.kafka.server there is package-info.java, I think it is public api, thus we should not modify the toString method

not really, the `pakcage-info.java` doesn't simply indicate "public" APIs. for example, this [package-info.java](https://github.com/apache/kafka/blob/ad031b99d312cdfe452965c841e22c1ab391956e/clients/src/main/java/org/apache/kafka/common/cache/package-info.java#L19) is used to announce this package is NOT public APIs.

Although it's not a public API, I prefer to maintain the existing format due to its long-standing presence.

chia7712 replied:
> In org.apache.kafka.server there is package-info.java

btw, in the server-common module the package `org.apache.kafka.server` has no `package-info.java`

m1a2st replied:
Sorry, there are two `TopicIdPartition` files, I trace the wrong directory :(

m1a2st replied:
> not really, the pakcage-info.java doesn't simply indicate "public" APIs. for example, this [package-info.java](https://github.com/apache/kafka/blob/ad031b99d312cdfe452965c841e22c1ab391956e/clients/src/main/java/org/apache/kafka/common/cache/package-info.java#L19) is used to announce this package is NOT public APIs.

Thanks for @chia7712 point out, it seems that we still have to verify it through the documents. [Kafka KIP](https://cwiki.apache.org/confluence/display/kafka/kafka+improvement+proposals)

```
All of the following are public interfaces that people build around:
    Binary log format
    The network protocol and api behavior
    Any class for which the build generates Javadoc.  This includes (but is not limited to; look for javadoc {} sections in build.gradle for the complete list):
        org/apache/kafka/common/serialization
        org/apache/kafka/common
        org/apache/kafka/common/errors
        org/apache/kafka/clients/producer
        org/apache/kafka/clients/consumer (eventually, once stable)
    Configuration, especially client configuration
    Monitoring
    Command line tools and arguments
```

clolov replied:
Okay, so the summary is the following and please correct me if I am wrong - we are sticking to the toString format because it has been like so for a while and not because this is a public-facing API requiring a KIP to change it. If this is the case, I am happy for this to go through!

frankvicky replied:
I think yes 😄

frankvicky commented on server/src/main/java/org/apache/kafka/server/Assignment.java:
```diff
@@ -96,23 +57,6 @@ boolean valid(int nodeId, MetadataImage image) {
-            successCallback);
-    }
-
     @Override
     public String toString() {
         StringBuilder bld = new StringBuilder();
```
Same reason: https://github.com/apache/kafka/pull/18783/files#r1938893604

m1a2st reviewed (approved):
Thanks for this patch, LGTM

clolov reviewed (approved):
Thank you!

clolov commented:
Heya @chia7712, are we cherry-picking these changes for the 4.0 release?

chia7712 commented:
> are we cherry-picking these changes for the 4.0 release?

This is unrelated to bug fix, so I think it is unnecessary to backport to 4.0 :)

ijuma commented:
I asked this in another similar PR, but how are we verifying that it's safe to make these changes? Also, for critical classes like `TopicIdPartition`, we need to verify there is no perf regression when it comes to `equals`/`hashCode` as these classes are used in hot paths as HashMap keys.

To be clear, I'm not saying these changes are not good, but we need to do the due diligence since records have a specific behavior that may not be the same as the existing classes.

frankvicky commented:
Hi @ijuma 
Thanks for the reminder. It's indeed an issue that I should research first, especially for such underlying implementation and hot path related change. I will keep that in mind. 🙇🏼 

Back to the question, I have read [Java 17 SE spec](https://docs.oracle.com/javase/specs/jls/se17/html/jls-8.html#jls-8.10). Following is the statement:

> Equality of an instance a of a record class R with another instance b of the same record class at a record component c is determined as follows:
If the type of the record component c is a reference type, equality is determined as follows: if the value of the component field c of both a and b is the null reference then true is returned; if the value of the component field c of either a or b, but not both, is the null reference then false is returned; otherwise equality is determined by invoking the equals method on the value of the component field c of a, with an argument that is the value of the component field c of b.
If the type of the record component c is a primitive type T, equality is determined as if by invoking the static method compare of the wrapper class corresponding to T ([§5.1.7](https://docs.oracle.com/javase/specs/jls/se17/html/jls-5.html#jls-5.1.7)), with the first argument given by the value of the component field c of a, and the second argument given by the value of the component field c of b; if the method would return 0 then true is returned, otherwise false is returned.

According to the spec, the record's `equals` method behavior is basically the same as what we implemented before.
Although primitive types don't directly use the `==` operator, the result is effectively the same. This can be seen in the `Integer.compare` implementation.

https://github.com/apache/kafka/blob/c215b8b645edf6769f4a8976edcf0cc380ded21b/server-common/src/main/java/org/apache/kafka/server/common/TopicIdPartition.java#L49-L54

**Integer.compare:**
```java
    public static int compare(int x, int y) {
        return (x < y) ? -1 : ((x == y) ? 0 : 1);
    }
```

> A method public final int hashCode() that returns a hash code value derived from the hash code values at every record component of R.
The hash code value of an instance a of a record class at a record component c is as follows:
If the type of the record component c is a reference type, then the hash code value is determined as if by invoking the hashCode method on the value of the component field c of a.

According to the spec, the behavior of the record's `hashCode` method is basically the same as what we implemented before:
https://github.com/apache/kafka/blob/c215b8b645edf6769f4a8976edcf0cc380ded21b/server-common/src/main/java/org/apache/kafka/server/common/TopicIdPartition.java#L56-L59


Therefore, the changes to `record` should be behavior-compatible. As for performance, since the underlying implementation remains essentially the same, there should not be significant performance differences.  🙇🏼

ijuma commented:
@TaiJuWu I know about the spec, but the actual implementation details matter. So, you would want to write a benchmark and confirm - not simply rely on reading the specification.

chia7712 commented:
> you would want to write a benchmark and confirm

@frankvicky Since we have many JIRA tickets related to rewriting code with record classes, it would be great to have a JMH benchmark to compare both equals and hashCode performance.  This benchmark would also give us more confidence when merging other related PRs.

frankvicky commented:
I see.
I will prepare one 🙇🏼

ijuma commented:
Yes, I think a benchmark for `equals`/`hashCode` for `TopicIdPartition` would be a good one since the original implementation is pretty simple and if the generated implementation has additional fixed overhead, that would be observed.
